### ISSUE TITLE
Potential Issue with Optimizer Parameter Binding in DDP Training

### ISSUE THREAD

**Issue by wangjunjie-1** [facebookresearch/watermark-anything#19]
Hello, I'm reviewing the training code and I've noticed what appears to be a potential issue regarding how the optimizer is initialized. It seems to be built directly from the parameters of the `embedder `and `extractor `submodules, rather than from the DDP-wrapped model.

The relevant code is as follows:
`
optimizer = uoptim.build_optimizer(
    model_params=list(embedder.parameters()) + list(extractor.parameters()),
)
`
My understanding is that, according to PyTorch's DDP best practices, the optimizer should be initialized with the parameters of the `ddp_model `to ensure it uses the synchronized, averaged gradients during the optimizer.step() call.

I would like to ask if this parameter binding method could potentially affect the correctness of DDP training. I might be missing some context; is there a special mechanism that handles this within the `uoptim.build_optimizer` function, or does some other part of the code ensure that the gradients are properly synchronized and applied?

Thank you for your time and clarification!

**Comment by pierrefdz** [comment]
Thanks for the careful read. In this code, initializing the optimizer from the embedder/extractor submodules is correct for DDP.

DDP synchronizes gradients during loss.backward() on the DDP-wrapped model (we call train_one_epoch with wam_ddp), and the Parameter tensors in embedder/extractor are the exact same objects now owned by wam and then wrapped by DDP. Optimizer.step() operates on those same tensors, whose grads were allreduced by DDP. PyTorch allows creating the optimizer before or after wrapping with DDP as long as the same Parameter objects are used.

uoptim.build_optimizer doesn’t do anything special for DDP; it just builds the requested torch optimizer.
